pandas/core/internals/construction.py

```diff
@@ -300,8 +300,10 @@ def extract_index(data):
             raise ValueError('If using all scalar values, you must pass'
                              ' an index')
 
-        if have_series or have_dicts:
+        if have_series:
             index = _union_indexes(indexes)
+        elif have_dicts:
+            index = _union_indexes(indexes, sort=False)
```

Comment:
So I suppose this is the key one. IIUC, we *want* to sort for Python=3.5. So this should be

```python
index = _union_indexes(indexes, sort=not compat.PY36)
```

That will be True for python 3.5 and False for 3.6 and newer.

You'll need to `import pandas.compat` at the top.

Reply:
I haven't checked yet if you need to inspect the contents of the dicts to see if they're ordered or not. *Hopefully* you won't have to.

Reply:
OK. I'll take care of this later.

Reply:
Instead of 
> ```python
> index = _union_indexes(indexes, sort=not compat.PY36)
> ```
changed to 
```python
index = _union_indexes(indexes, sort=not (compat.PY36 or have_ordered)) 
```
to prevent sorting a OrderedDict in Python3.5